src/legacy/core_plugins/kibana/public/dashboard/np_ready/dashboard_app_controller.tsx

```diff
@@ -713,8 +715,44 @@ export class DashboardAppController {
         });
     }
 
+    const hasUrlParam = (param: string): boolean =>
+      param in parse(location.hash.slice(1), true).query;
+
+    const displayIfUrlParam = (param: string): boolean =>
+      hasUrlParam(param) && !dashboardStateManager.getFullScreenMode();
+
+    const topNavUpdate = (): void => {
+      $scope.topNavMenu = $scope.showTopNavMenu()
+        ? getTopNavConfig(
+            dashboardStateManager.getViewMode(),
+            navActions,
+            dashboardConfig.getHideWriteControls()
+          )
+        : null;
+    };
+
+    $scope.$watch(
+      (): boolean => $scope.showTopNavMenu(),
+      (): void => topNavUpdate()
+    );
+
+    $scope.showTopNav = () => $scope.isVisible || $scope.showSearchBar();
+
+    $scope.showTopNavMenu = () => $scope.isVisible || displayIfUrlParam('show-top-nav-menu');
+
+    $scope.showSearchBar = () =>
+      $scope.isVisible || $scope.showQueryBar() || $scope.showFilterBar();
+
+    $scope.showQueryBar = () =>
+      $scope.isVisible || $scope.showQueryInput() || $scope.showDatePicker();
+
+    $scope.showQueryInput = () => $scope.isVisible || displayIfUrlParam('show-query-input');
+
+    $scope.showDatePicker = () => $scope.isVisible || displayIfUrlParam('show-date-picker');
```

Comment:
I've used functions for all of these, similar to `showFilterBar`, because this made it easier to test while developing as I could update the URL and see the changes immediately. My thoughts are that this is better than setting up watchers similar to `showSaveQuery`, and also better than only checking the value on page load in case users want to enable certain functionality dynamically.

Reply:
It's a bit repetitive, but I'm ok with this as this code will change (somewhat) soon 😬